src/client/metric.rs

```diff
@@ -0,0 +1,528 @@
+    pub time_dim: i8,
+    /// Count dimension (uses least-significant 4 bits only)
+    pub count_dim: i8
+}
+
+lazy_static! {
+    static ref SPACE_UNIT: Unit = {
+        let mut unit = Unit::empty();
+        unit.space_dim = 1;
+        unit
+    };
+
+    static ref TIME_UNIT: Unit = {
+        let mut unit = Unit::empty();
+        unit.time_dim = 1;
+        unit
+    };
+
+    static ref COUNT_UNIT: Unit = {
+        let mut unit = Unit::empty();
+        unit.count_dim = 1;
+        unit
+    };
+}
+
+impl Unit {
+    /// Returns an empty unit with all dimensions set to `0`
+    /// and all scales set to an undefined variant
+    pub fn empty() -> Self {
+        Unit {
+            space_scale: SpaceScale::Byte,
+            time_scale: TimeScale::NSec,
+            count_scale: CountScale::One,
+            space_dim: 0,
+            time_dim: 0,
+            count_dim: 0
+        }
+    }
+
+    /// Returns a unit with space dimension `1` and given
+    /// space scale
+    pub fn space(space_scale: SpaceScale) -> Self {
+        let mut unit = SPACE_UNIT.clone();
+        unit.space_scale = space_scale;
+        unit
+    }
+
+    /// Returns a unit with time dimension `1` and given
+    /// time scale
+    pub fn time(time_scale: TimeScale) -> Self {
+        let mut unit = TIME_UNIT.clone();
+        unit.time_scale = time_scale;
+        unit
+    }
+
+    /// Returns a unit with count dimension `1` and given
+    /// count scale
+    pub fn count() -> Self {
+        COUNT_UNIT.clone()
+    }
+
+    /*
+        pmapi representation of a unit. below, 31 refers to MSB
+        bits 31 - 28 : space dim   (signed)
+             27 - 24 : time dim    (signed)
+             23 - 20 : count dim   (signed)
+             19 - 16 : space scale (unsigned)
+             15 - 12 : time scale  (unsigned)
+             11 - 8  : count scale (unsigned)
+              7 - 0  : zero pad
+    */
+    fn pmapi_repr(&self) -> u32 {
+        let mut repr = 0;
+
+        repr |= ((self.space_dim as i32) & ((1 << 4) - 1)) << 28;
+        repr |= ((self.time_dim  as i32) & ((1 << 4) - 1)) << 24;
+        repr |= ((self.count_dim as i32) & ((1 << 4) - 1)) << 20;
+
+        repr |= (self.space_scale as i32) << 16;
+        repr |= (self.time_scale  as i32) << 12;
+        repr |= (self.count_scale as i32) << 8;
+
+        repr as u32
+    }
+}
+
+#[derive(Copy, Clone)]
+/// Semantic for a Metric
+pub enum Semantics {
+    /// Counter
+    Counter  = 1,
+    /// Instant
+    Instant  = 3,
+    /// Discrete
+    Discrete = 4
+}
+
+/// Singleton metric
+pub struct Metric<T> {
+    name: String,
+    item: u32,
+    sem: Semantics,
+    indom: u32,
+    unit: u32,
+    shorthelp: String,
+    longhelp: String,
+    val: T,
+    mmap_view: MmapViewSync
+}
+
+lazy_static! {
+    static ref SCRATCH_VIEW: MmapViewSync = {
+        Mmap::anonymous(super::STRING_BLOCK_LEN as usize, Protection::ReadWrite).unwrap()
+            .into_view_sync()
```

Comment:
Then we'd have to do a runtime check for a valid `mmap_view` in `set_val`, and I wanted to reduce the internal branches in that particular function.